### DISCUSSION TITLE
Laravel Cashier (Stripe): You must provide at least one recurring price in `subscription` mode when using prices. while trying both subscription and one-time charge

### DISCUSSION THREAD

**Discussion by devhoussam1998** [laravel/framework#51319] [Q&A]
I have a product in Stripe with multiple prices, some recurring and a one-time price. (eg. $10/month, $100/year, $1,000 one-time)

For the recurring prices I use `$request->user()->newSubscription()` with `stripe_product`, and `stripe_price` it works well but when I try to handle the one-time price I get this error:

**You must provide at least one recurring price in `subscription` mode when using prices.**

What is the correct way to handle a recurring and one-time charges.

**SubscriptionCheckoutController.php**

```
<?php

namespace App\Http\Controllers;

use App\Models\SubscriptionPlan;
use Illuminate\Http\Request;

class SubscriptionCheckoutController extends Controller
{
    /**
     * Handle the incoming request.
     */
    public function __invoke(Request $request, string $priceId)
    {
        // Find the subscription plan by the given price ID
        $plan = SubscriptionPlan::where('stripe_price', $priceId)->firstOrFail();

        // Create a new subscription builder for the user
        $subscriptionBuilder = $request->user()->newSubscription($plan->stripe_product, $plan->stripe_price);

        // Set trial days if they exist
        if ($plan->trial_days !== null) {
            $subscriptionBuilder->trialDays($plan->trial_days);
        }

        // Allow promotion codes if enabled
        if ($plan->allow_promotion_codes) {
            $subscriptionBuilder->allowPromotionCodes();
        }

        // Perform checkout with specified URLs
        return $subscriptionBuilder->checkout([
            'success_url' => route('success-url'),
            'cancel_url' => route('cancel-url'),
        ]);
    }
}
```

**Comment by joy2362** [comment]
are you want to achive with one time payment is it something like life time payment ? correct me if i'm wrong

**Reply by devhoussam1998** [reply]
I have two type: eithere with subscribtion mothly or yearly or with lifetime payment

**Reply by joy2362** [reply]
ok I got it i faced same issue before and I'd contact with stripe support and they give me a salutation which is create a recurring price by set 

```
"interval": "year",
"interval_count": 999,

```
it will make your subscription life time plan stripe not support life time plan directly so you can try this way hope it solve your problem 

another solution i'm not sure if it works or not is you can take single payment and add that user with a free plan like create a price with 0 per month or year whatever you like but issue is it will generate invoice for that user every month 

personally i think 1st solution better as i dont need extra logic for subscription

**Reply by devhoussam1998** [reply]
Thanks @joy2362 i will give it a try now and let you know

**Reply by joy2362** [reply]
1st option?

**Reply by devhoussam1998** [reply]
@joy2362 I try to update the price from stripe dashboard in **Billing period** I use custom but I get the error message:

**Billing period must be between 1 day and 3 years**

<img width="589" alt="Screenshot 2024-05-07 at 11 46 17" src="https://github.com/laravel/framework/assets/121255432/ca5b31bf-55d0-481b-9db4-946708d60a90">

**Reply by joy2362** [reply]
sorry about that here is the full conversion that i made with stripe 

![WhatsApp Image 2024-05-07 at 17 02 06_a84dcc3a](https://github.com/laravel/framework/assets/49846472/c90b2213-2458-4024-92e0-61e4c825fd97)
![WhatsApp Image 2024-05-07 at 17 02 06_8cb9cb4a](https://github.com/laravel/framework/assets/49846472/d5c19d19-43f1-43b5-9b27-7cffa2b396b6)

**Reply by devhoussam1998** [reply]
so the idea is to create **One-off** price from stripe dashboard I already do that.
next i check the laravel cashier docs exactly in [Single Charges](https://laravel.com/docs/11.x/billing#single-charges)

I made this example but I get this error message once a user logged in with no subscription and try to select **lifetime** price.

**Error Exception is**:

**You must provide at least one recurring price in subscription mode when using prices.**

```
SubscriptionCheckoutController.php

class SubscriptionCheckoutController extends Controller
{
    /**
     * Handle the incoming request.
     */
    public function __invoke(Request $request, string $priceId)
    {
        // Find the subscription plan by the given price ID
        $plan = SubscriptionPlan::where('stripe_price', $priceId)->firstOrFail();

        if ($plan->periodicity_type !== PeriodicityType::Forever) {
            // Create a new subscription builder for the user
            $subscriptionBuilder = $request->user()->newSubscription($plan->stripe_product, $plan->stripe_price);

            // Set trial days if they exist
            if ($plan->trial_days !== null) {
                $subscriptionBuilder->trialDays($plan->trial_days);
            }

            // Allow promotion codes if enabled
            if ($plan->allow_promotion_codes) {
                $subscriptionBuilder->allowPromotionCodes();
            }

            // Handle successful subscription
            return $subscriptionBuilder->checkout([
                'success_url' => route('home'),
                'cancel_url' => route('billing'),
            ]);
        } else {
            // Handle one-time payment for "Forever" plan
            if ($plan->periodicity_type === PeriodicityType::Forever) {
                // Skip subscription creation and directly charge the user
                $singleCharge = $request->user()->charge(
                    $plan->price,
                    $request->paymentMethodId
                );

                // Handle successful charge
                return redirect()->route('home');
            }
        }
    }
}
```
